Add Public API support for function pointer types

333fred opened the pull request:
Implements and documents the public API access points for function pointer types.

333fred commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1991,6 +1991,44 @@ internal PointerTypeSymbol CreatePointerTypeSymbol(TypeSymbol elementType, Nulla
+                throw new ArgumentException(string.Format(CSharpResources.NotSameNumberParameterTypesAndRefKinds, parameterTypes.Length, parameterRefKinds.Length));
+            }
+
+            if (returnRefKind == RefKind.Out)
```
I'm unsure how much validation we should be doing here. For example, should we be validating that `void` is not paired with a refkind? Or that `void` is not provided as a parameter type?

333fred commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1991,6 +1991,42 @@ internal PointerTypeSymbol CreatePointerTypeSymbol(TypeSymbol elementType, Nulla
+                throw new ArgumentException(string.Format(CSharpResources.NotSameNumberParameterTypesAndRefKinds, parameterTypes.Length, parameterRefKinds.Length));
+            }
+
+            if (returnRefKind == RefKind.Out)
```
I'm unsure what level of validation we should be doing here. For example, should we check that `void` isn't passed for a parameter type, and that `void` returns always have `RefKind.None`?

AlekseyTs replied:
> I'm unsure what level of validation we should be doing here. For example, should we check that void isn't passed for a parameter type, and that void returns always have RefKind.None?

I do not think we do this kind of validation in other similar scenarios, like constructing a generic type, anonymous type, etc.

---
In reply to: [428297776](https://github.com/dotnet/roslyn/pull/44436#discussion_r428297776) [](ancestors = 428297776)

AlekseyTs replied:
That said, I think it makes sense to keep the check for ```returnRefKind```.

---
In reply to: [428846682](https://github.com/dotnet/roslyn/pull/44436#discussion_r428846682) [](ancestors = 428846682,428297776)

333fred replied:
Read my mind about my next follow up question, thanks!

---
In reply to: [428848084](https://github.com/dotnet/roslyn/pull/44436#discussion_r428848084) [](ancestors = 428848084,428846682,428297776)

333fred commented:
@AlekseyTs @dotnet/roslyn-compiler for review. /cc @jasonmalinowski @CyrusNajmabadi, I've added the `CreateFunctionPointerTypeSymbol` api in this change. It doesn't have calling convention yet, as we don't have a public API for that yet either.

jasonmalinowski commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1991,6 +1991,42 @@ internal PointerTypeSymbol CreatePointerTypeSymbol(TypeSymbol elementType, Nulla
+
+            if (returnRefKind == RefKind.Out)
+            {
+                //'RefKind.Out' is not a valid ref kind for a return type.
+                throw new ArgumentException(CSharpResources.OutIsNotValidForReturn);
```
If somebody writes `delegate*<out int>` today, is the existing symbol going to still mark that as out or is it not even going to make it into the symbol model? My concern being this is fine as long as it's blocked from ever appearing in any symbol anywhere, not just through callers of this API.

333fred replied:
I'm not sure what the public api would do, actually, it's a good test to add.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -939,6 +939,13 @@ public IPointerTypeSymbol CreatePointerTypeSymbol(ITypeSymbol pointedAtType)
 
         protected abstract IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol elementType);
 
+        public IFunctionPointerTypeSymbol CreateFunctionPointerTypeSymbol(ITypeSymbol returnType, RefKind returnRefKind, ImmutableArray<ITypeSymbol> parameterTypes, ImmutableArray<RefKind> parameterRefKinds)
```
>CreateFunctionPointerTypeSymbol [](start = 42, length = 31)

Please add a doc comment, including for parameters, what values are allowed, etc.

CyrusNajmabadi replied:
Shape seems reasonable for IDE purposes.  Note sure how calling convention is represented though...

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -299,6 +299,10 @@ internal IOperation CreateInternal(BoundNode boundNode)
                     return CreateBoundSwitchExpressionArmOperation((BoundSwitchExpressionArm)boundNode);
                 case BoundKind.ObjectOrCollectionValuePlaceholder:
                     return CreateCollectionValuePlaceholderOperation((BoundObjectOrCollectionValuePlaceholder)boundNode);
+                case BoundKind.FunctionPointerInvocation:
```
>case BoundKind.FunctionPointerInvocation: [](start = 16, length = 41)

Please make sure to run all tests with IOperation test hook enabled and confirm that there are no failures related to the "Function pointers" feature.

333fred replied:
I've run them locally and gone through the failures. Nothing related to this feature.

---
In reply to: [428783559](https://github.com/dotnet/roslyn/pull/44436#discussion_r428783559) [](ancestors = 428783559)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -434,6 +438,44 @@ private IOperation CreateBoundCallOperation(BoundCall boundCall)
 
+        private IOperation CreateBoundFunctionPointerInvocationOperation(BoundFunctionPointerInvocation boundFunctionPointerInvocation)
+        {
+            if (boundFunctionPointerInvocation.HasErrors)
```
>.HasErrors [](start = 46, length = 10)

This is a very imprecise test. Is it possible to make it more specific, check ```ResultKind```, for example?

333fred replied:
Yeah, I think we could. I hadn't added `ResultKind` yet when I implemented this part, I think it's probably a better fit.

---
In reply to: [428786732](https://github.com/dotnet/roslyn/pull/44436#discussion_r428786732) [](ancestors = 428786732)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -434,6 +438,44 @@ private IOperation CreateBoundCallOperation(BoundCall boundCall)
+            return new CSharpLazyInvocationOperation(
+                this,
+                boundFunctionPointerInvocation,
+                boundFunctionPointerInvocation.FunctionPointer.Signature.GetPublicSymbol(),
```
>boundFunctionPointerInvocation.FunctionPointer.Signature.GetPublicSymbol() [](start = 16, length = 74)

This looks suspicious. I am not sure it is the right thing to pretend that this is the method that is getting invoked.

AlekseyTs replied:
I would say we need a special IOperation node to represent a function pointer invocation.

---
In reply to: [428789347](https://github.com/dotnet/roslyn/pull/44436#discussion_r428789347) [](ancestors = 428789347)

333fred replied:
I don't think I agree. Function pointer invocations are one of the `callX` Il instructions, and I feel like it should be represented as a standard invocation. Semantically, it uses the same parameter conversion rules as a standard invocation, the same invocation order rules... I'm also having trouble seeing when you'd want to analyze invocations and not also look for function pointer invocations. @cyrusnajmabadi, do you want to weigh as well here?

---
In reply to: [428790455](https://github.com/dotnet/roslyn/pull/44436#discussion_r428790455) [](ancestors = 428790455,428789347)

AlekseyTs replied:
> I'm also having trouble seeing when you'd want to analyze invocations and not also look for function pointer invocations.

I don't believe this is relevant. We don't bundle things together just because people might want to look at them.

---
In reply to: [428829053](https://github.com/dotnet/roslyn/pull/44436#discussion_r428829053) [](ancestors = 428829053,428790455,428789347)

AlekseyTs replied:
I believe the two first properties are not applicable for function pointer invocations. The values that we are going to return are going to be misleading and simply will not match the API design. FP signature is not a method definition and it is not a member of the pointer instance.
```
    public interface IInvocationOperation : IOperation
    {
        /// <summary>
        /// Method to be invoked.
        /// </summary>
        IMethodSymbol TargetMethod { get; }
        /// <summary>
        /// 'This' or 'Me' instance to be supplied to the method, or null if the method is static.
        /// </summary>
        IOperation Instance { get; }
```

---
In reply to: [428832354](https://github.com/dotnet/roslyn/pull/44436#discussion_r428832354) [](ancestors = 428832354,428829053,428790455,428789347)

AlekseyTs replied:
The third property is also not applicable and will be misleading no matter what value we return:
```
        /// <summary>
        /// True if the invocation uses a virtual mechanism, and false otherwise.
        /// </summary>
        bool IsVirtual { get; }
```

---
In reply to: [428837184](https://github.com/dotnet/roslyn/pull/44436#discussion_r428837184) [](ancestors = 428837184,428832354,428829053,428790455,428789347)

333fred replied:
I'm going to make this a `NoneOperation` for now, and we can figure out a type shape later.

---
In reply to: [428841690](https://github.com/dotnet/roslyn/pull/44436#discussion_r428841690) [](ancestors = 428841690,428837184,428832354,428829053,428790455,428789347)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -434,6 +438,44 @@ private IOperation CreateBoundCallOperation(BoundCall boundCall)
+                this,
+                boundFunctionPointerInvocation,
+                boundFunctionPointerInvocation.FunctionPointer.Signature.GetPublicSymbol(),
+                isVirtual: false,
```
>isVirtual: false, [](start = 16, length = 17)

We don't know that, right?

333fred replied:
We do. Function pointers do not `callvirt`. The method is not dynamically resolved based on the vtable.

---
In reply to: [428790957](https://github.com/dotnet/roslyn/pull/44436#discussion_r428790957) [](ancestors = 428790957)

AlekseyTs replied:
> The method is not dynamically resolved based on the vtable.

We don't know that. This is part of the pointer. 

---
In reply to: [428829524](https://github.com/dotnet/roslyn/pull/44436#discussion_r428829524) [](ancestors = 428829524,428790957)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -434,6 +438,44 @@ private IOperation CreateBoundCallOperation(BoundCall boundCall)
+                _semanticModel,
+                boundFunctionPointerInvocation.Syntax,
+                boundFunctionPointerInvocation.Type.GetPublicSymbol(),
+                constantValue: default,
```
>default [](start = 31, length = 7)

It looks like for an error case we do ```ConvertToOptional(boundFunctionPointerInvocation.ConstantValue)```, but not here. Feels inconsistent.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -900,6 +942,25 @@ private IOperation CreateBoundConversionOperation(BoundConversion boundConversio
             BoundExpression boundOperand = boundConversion.Operand;
             if (boundConversion.ConversionKind == CSharp.ConversionKind.MethodGroup)
             {
+                if (boundConversion is { Type: FunctionPointerTypeSymbol { Signature: var signature }, Syntax: PrefixUnaryExpressionSyntax addressOfExpression })
```
>PrefixUnaryExpressionSyntax [](start = 111, length = 27)

How do we know this is an address of? Besides, is it really important what syntax we have here? We have a bound node, semantics of bound nodes doesn't depend on associated syntax node. The translation should be based on semantics. Is this a method group conversion to a function pointer? Then it is not a delegate creation regardless of the syntax.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -900,6 +942,25 @@ private IOperation CreateBoundConversionOperation(BoundConversion boundConversio
+                {
+                    return new AddressOfOperation(
+                        new MethodReferenceOperation(
+                            signature.GetPublicSymbol(),
```
>signature.GetPublicSymbol() [](start = 28, length = 27)

This doesn't look right. This should be the method the pointer will point to, I think.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -900,6 +942,25 @@ private IOperation CreateBoundConversionOperation(BoundConversion boundConversio
+                        new MethodReferenceOperation(
+                            signature.GetPublicSymbol(),
+                            isVirtual: false,
+                            instance: null,
```
>instance: null [](start = 28, length = 14)

Is this null even in error cases?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -900,6 +942,25 @@ private IOperation CreateBoundConversionOperation(BoundConversion boundConversio
+                        syntax: addressOfExpression,
+                        type: boundConversion.Type.GetPublicSymbol(),
+                        constantValue: default,
+                        isImplicit: false);
```
>isImplicit: false [](start = 24, length = 17)

How do we know that?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -900,6 +942,25 @@ private IOperation CreateBoundConversionOperation(BoundConversion boundConversio
+                if (boundConversion is { Type: FunctionPointerTypeSymbol { Signature: var signature }, Syntax: PrefixUnaryExpressionSyntax addressOfExpression })
+                {
+                    return new AddressOfOperation(
+                        new MethodReferenceOperation(
```
>MethodReferenceOperation [](start = 28, length = 24)

It looks like we are dropping ```boundConversion.Operand``` on the floor.

AlekseyTs replied:
>MethodReferenceOperation [](start = 28, length = 24)

It feels like we should be using ```CreateBoundMethodGroupSingleMethodOperation``` or something similar to create this node

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:
```diff
@@ -243,75 +243,68 @@ internal ImmutableArray<IArgumentOperation> DeriveArguments(BoundNode containing
 
         internal ImmutableArray<IArgumentOperation> DeriveArguments(BoundNode containingExpression)
         {
-            switch (containingExpression.Kind)
```
>switch (containingExpression.Kind) [](start = 12, length = 34)

Please undo the refactoring. I believe this big switch is easier to follow when it uses a statement form.

AlekseyTs replied:
Please keep changes to the necessary minimum to implement the feature. Addition of a ne case shouldn't trigger rewrite of all existing cases.

---
In reply to: [428807548](https://github.com/dotnet/roslyn/pull/44436#discussion_r428807548) [](ancestors = 428807548)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs:
```diff
@@ -817,18 +822,14 @@ internal sealed class CSharpLazyInvocationOperation : LazyInvocationOperation
         protected override IOperation CreateInstance()
         {
-            BoundExpression receiver;
-            switch (_invocableExpression)
```
>switch (_invocableExpression) [](start = 12, length = 29)

Please revert refactoring, simply add new case that you now want to handle. It should be obvious to a reviewer that other cases are not touched and the best way to achieve this is to not touch them.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -10,6 +10,7 @@
 using System.Linq;
 using System.Reflection;
 using Microsoft.Cci;
+using Microsoft.CodeAnalysis;
```
Is this necessary? It looks like we are inside this namespace already.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -92,6 +93,23 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
                 suppressUseSiteDiagnostics);
         }
 
+        internal static FunctionPointerMethodSymbol CreateFromParts(
```
>CreateFromParts [](start = 52, length = 15)

Consider adding a comment that this helper can be used only in scenarios that are not interested in reporting any diagnostics. And probably the same comment for the new constructor.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:
```diff
@@ -25,6 +25,14 @@ public static FunctionPointerTypeSymbol CreateFromSource(FunctionPointerTypeSynt
                     basesBeingResolved,
                     suppressUseSiteDiagnostics));
 
+        public static FunctionPointerTypeSymbol CreateFromParts(
```
>CreateFromParts [](start = 48, length = 15)

It would be good to have similar comment on this API.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1991,6 +1991,42 @@ internal PointerTypeSymbol CreatePointerTypeSymbol(TypeSymbol elementType, Nulla
+
+            if (parameterTypes.IsDefault || parameterTypes.Any(param => param.IsDefault))
+            {
+                throw new ArgumentNullException(nameof(parameterTypes));
```
>ArgumentNullException [](start = 26, length = 21)

This exception isn't really applicable to the second condition.

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6154,4 +6154,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>The calling convention of '{0}' is not supported by the language.</value>
   </data>
+  <data name="NotSameNumberParameterTypesAndRefKinds" xml:space="preserve">
+    <value>Given {0} parameter types and {1} parameter ref kinds. These must be the same.</value>
```
>Given {0} parameter types and {1} parameter ref kinds. These must be the same. [](start = 11, length = 78)

Consider making it explicit that both sentences talk about count of things.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1991,6 +1991,42 @@ internal PointerTypeSymbol CreatePointerTypeSymbol(TypeSymbol elementType, Nulla
             return new PointerTypeSymbol(TypeWithAnnotations.Create(elementType, elementNullableAnnotation));
         }
 
+        internal FunctionPointerTypeSymbol CreateFunctionPointerTypeSymbol(
```
>CreateFunctionPointerTypeSymbol [](start = 43, length = 31)

Is there more than one call site? Is it worth duplicating validation logic?

AlekseyTs replied:
Can we make it private?

---
In reply to: [428852062](https://github.com/dotnet/roslyn/pull/44436#discussion_r428852062) [](ancestors = 428852062)

333fred replied:
Yeah, we could. I was following the precedent from the other internal create methods. I could even go as far as removing this API and just leaving `Common`, and if we ever need to add it we can break it out again.

---
In reply to: [428894906](https://github.com/dotnet/roslyn/pull/44436#discussion_r428894906) [](ancestors = 428894906,428852062)

AlekseyTs replied:
> I could even go as far as removing this API and just leaving Common, and if we ever need to add it we can break it out again.

Sounds reasonable

---
In reply to: [428902442](https://github.com/dotnet/roslyn/pull/44436#discussion_r428902442) [](ancestors = 428902442,428894906,428852062)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -3223,6 +3223,14 @@ public ILocalSymbol GetDeclaredSymbol(CatchDeclarationSyntax catchDeclaration, C
+                case BoundKind.FunctionPointerInvocation:
+                    {
+                        var invocation = (BoundFunctionPointerInvocation)boundNode;
+                        symbols = ImmutableArray.Create<Symbol>(invocation.FunctionPointer.Signature);
```
>invocation.FunctionPointer.Signature [](start = 64, length = 36)

This doesn't feel right. I think the symbol should be the pointer.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -3290,6 +3298,10 @@ public ILocalSymbol GetDeclaredSymbol(CatchDeclarationSyntax catchDeclaration, C
                         }
+                        else if (conversion.ConversionKind == ConversionKind.NoConversion && conversion.Operand is BoundUnconvertedAddressOfOperator unconvertedAddressOf)
+                        {
+                            symbols = GetMethodGroupSemanticSymbols(unconvertedAddressOf.Operand, boundNodeForSyntacticParent, binderOpt, out resultKind, out isDynamic, out memberGroup);
```
>symbols = GetMethodGroupSemanticSymbols(unconvertedAddressOf.Operand, boundNodeForSyntacticParent, binderOpt, out resultKind, out isDynamic, out memberGroup); [](start = 28, length = 158)

It is not obvious what scenario we are trying to address here. Could you please elaborate?

333fred replied:
If there wasn't a conversion, the default behavior is get the `ExpressionSymbol`, if there is one. For unconverted address of operators, this isn't correct: we want to get all the symbols from the method group that underlies the address of. I can add this as a comment in source.

---
In reply to: [428854021](https://github.com/dotnet/roslyn/pull/44436#discussion_r428854021) [](ancestors = 428854021)

AlekseyTs replied:
Examples might help. Could you point to specific unit-test(s)?

---
In reply to: [428885643](https://github.com/dotnet/roslyn/pull/44436#discussion_r428885643) [](ancestors = 428885643,428854021)

333fred replied:
`PublicApi_GetSymbolInfo`, where the conversion is incorrect. Because the conversion fails, we don't end up creating a method group conversion and we have a `BoundUnconvertedAddressOfOperator` left. I'll add an additional tests to further exercise this path around attempting to get the address of a missing method.

---
In reply to: [428891527](https://github.com/dotnet/roslyn/pull/44436#discussion_r428891527) [](ancestors = 428891527,428885643,428854021)

333fred replied:
I could also be persuaded that the underlying identifier syntax should be the thing that returns the method symbols, and not the whole `&M` syntax. I'm not aware of any particular api precedence we have in this area, so your opinion would be appreciated.

---
In reply to: [428896674](https://github.com/dotnet/roslyn/pull/44436#discussion_r428896674) [](ancestors = 428896674,428891527,428885643,428854021)

AlekseyTs replied:
> I could also be persuaded that the underlying identifier syntax should be the thing that returns the method symbols, and not the whole &M syntax. I'm not aware of any particular api precedence we have in this area, so your opinion would be appreciated.

I think it is reasonable to return a symbol for ```&M``` syntax. However, we should still decide what, if anything, should be returned for the ```M```. We also have an API called GetMemberGroup (or something similar). I think it should be applicable to ```M``` syntax. Consider checking what we do right now for method group to delegate conversion, it might be good to be consistent in some way even though we have an extra syntax node involved. A regular invocation is another scenario to check.

---
In reply to: [429357637](https://github.com/dotnet/roslyn/pull/44436#discussion_r429357637) [](ancestors = 429357637,428896674,428891527,428885643,428854021)

333fred replied:
This is now handled correctly. My existing solution worked for `GetSymbolInfo`, but ran into issues with `GetMemberGroup`. I've updated it to handle `BoundUnconvertedAddressOfOperator` directly, which removes the need for this special handling here. The only adjustments we need to make are around successful method-group-to-function-pointer conversions.

---
In reply to: [429412739](https://github.com/dotnet/roslyn/pull/44436#discussion_r429412739) [](ancestors = 429412739,429357637,428896674,428891527,428885643,428854021)

CyrusNajmabadi commented on src/Compilers/Core/Portable/Symbols/IFunctionPointerTypeSymbol.cs:
```diff
@@ -5,10 +5,20 @@
+        /// <summary>
+        /// Gets the signature of the function pointed to by an instance of the function pointer type.
+        /// </summary>
         public IMethodSymbol Signature { get; }
```
how is calling convention represented?

333fred replied:
See the comment above: it is not today.

---
In reply to: [428879107](https://github.com/dotnet/roslyn/pull/44436#discussion_r428879107) [](ancestors = 428879107)

AlekseyTs commented on src/Compilers/Test/Utilities/CSharp/FunctionPointerUtilities.cs:
```diff
@@ -128,6 +130,42 @@ static void verifyParameter(ParameterSymbol symbol, MethodSymbol containing)
+                Assert.Equal(typeInfo.Type, typeInfo.ConvertedType, SymbolEqualityComparer.IncludeNullability);
+            }
+
+            CommonVerifyFunctionPointer((FunctionPointerTypeSymbol)((CSharp.Symbols.PublicModel.FunctionPointerTypeSymbol)typeSymbol).UnderlyingTypeSymbol);
```
>(FunctionPointerTypeSymbol)((CSharp.Symbols.PublicModel.FunctionPointerTypeSymbol)typeSymbol).UnderlyingTypeSymbol [](start = 40, length = 114)

I think we have GetSymbol (or something like that) helpers to do this. Consider adding a helper for FunctionPointerTypeSymbol if necessary.

333fred replied:
Good to know, thanks.

---
In reply to: [428899939](https://github.com/dotnet/roslyn/pull/44436#discussion_r428899939) [](ancestors = 428899939)

AlekseyTs commented on src/Compilers/Test/Utilities/CSharp/FunctionPointerUtilities.cs:
```diff
@@ -128,6 +130,42 @@ static void verifyParameter(ParameterSymbol symbol, MethodSymbol containing)
+
+            Assert.Equal(paramVerifiers.Length, typeSymbol.Signature.Parameters.Length);
+
+            foreach (var (verifier, parameter) in paramVerifiers.Zip(typeSymbol.Signature.Parameters, (v, p) => (v, p)))
```
>foreach (var (verifier, parameter) in paramVerifiers.Zip(typeSymbol.Signature.Parameters, (v, p) => (v, p))) [](start = 12, length = 108)

Why so complicated? Can't we just do a for based on index?

333fred replied:
I don't particularly find `for` loops where I have to manually keep track of indexes simpler than a `Zip` monda.

---
In reply to: [428900604](https://github.com/dotnet/roslyn/pull/44436#discussion_r428900604) [](ancestors = 428900604)

AlekseyTs commented on src/Compilers/Test/Utilities/CSharp/FunctionPointerUtilities.cs:
```diff
@@ -128,6 +130,42 @@ static void verifyParameter(ParameterSymbol symbol, MethodSymbol containing)
+            if (syntax is FunctionPointerTypeSyntax { Parameters: var parameterSyntaxes })
+            {
+                Assert.Equal(paramVerifiers.Length + 1, parameterSyntaxes.Count);
+                foreach (var (verifier, paramSyntax) in paramVerifiers.Append(returnTypeVerifier).Zip(parameterSyntaxes, (v, p) => (v, p)))
```
>foreach (var (verifier, paramSyntax) in paramVerifiers.Append(returnTypeVerifier).Zip(parameterSyntaxes, (v, p) => (v, p))) [](start = 16, length = 123)

Also looks unnecessary complicated.

AlekseyTs commented on src/Compilers/Test/Utilities/CSharp/FunctionPointerUtilities.cs:
```diff
@@ -128,6 +130,42 @@ static void verifyParameter(ParameterSymbol symbol, MethodSymbol containing)
+                {
+                    var type = model.GetTypeInfo(paramSyntax.Type!).Type!;
+                    Assert.NotNull(type);
+                    verifier(type);
```
>verifier [](start = 20, length = 8)

Why do we need to call verifier again? Would it be sufficient to assert that the type is equal to the one from the symbol? Same for ConvertedTo type.

AlekseyTs commented on src/Compilers/Test/Utilities/CSharp/FunctionPointerUtilities.cs:
```diff
@@ -128,6 +130,42 @@ static void verifyParameter(ParameterSymbol symbol, MethodSymbol containing)
             }
         }
 
+        public static void VerifyPublicFunctionPointerGetTypeInfo(SemanticModel model, SyntaxNode syntax, bool useConvertedType,
```
>VerifyPublicFunctionPointerGetTypeInfo [](start = 27, length = 38)

It feels like we would want to cover GetSymbolInfo here the same way as well.

AlekseyTs replied:
>syntax [](start = 98, length = 6)

I find this loosely typed helper extremely confusing and hard to follow at use-sites.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+            var syntaxTree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(syntaxTree);
+
+            var mDeclSyntax = syntaxTree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Skip(2).Single();
```
>mDeclSyntax [](start = 16, length = 11)

Please assert ToString for all syntax nodes of interest in order to eliminate all the guess work about what exactly we are looking at.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+            verifyStringReturn(mDeclSyntax.ParameterList.Parameters[0].Type!, useConvertedType: false);
+
+            var varDecl = mDeclSyntax.DescendantNodes().OfType<VariableDeclarationSyntax>().Single();
+            verifyVoidReturn(varDecl.Type, useConvertedType: false);
```
>verifyVoidReturn [](start = 12, length = 16)

General feedback. Extracted helpers certainly make code smaller, but they are doing this at the expense of clarity. Especially when different kinds of syntax nodes are fed to them. For example, a helper focused on a specific syntax node type would be much clearer. I am, as a reviewer, having a very hard time trying to figure out what do we test here and whether there are interesting missing scenarios.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+
+            var varDecl = mDeclSyntax.DescendantNodes().OfType<VariableDeclarationSyntax>().Single();
+            verifyVoidReturn(varDecl.Type, useConvertedType: false);
+            verifyVoidReturn(varDecl.Variables.Single().Initializer!.Value, useConvertedType: true);
```
>varDecl.Variables.Single().Initializer!.Value [](start = 29, length = 45)

I think we should also assert SemanticModel behavior for the operand of an address of. I am assuming we are testing an address of node here.

333fred replied:
`VerifyFunctionPointerSemanticInfo` will do this, asserting that it's the same expected symbols as the containing address of expression.

---
In reply to: [428913130](https://github.com/dotnet/roslyn/pull/44436#discussion_r428913130) [](ancestors = 428913130)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+
+            var assignment = mDeclSyntax.DescendantNodes().OfType<AssignmentExpressionSyntax>().Single();
+
+            verifyStringReturn(assignment, useConvertedType: false);
```
>verifyStringReturn [](start = 12, length = 18)

I do not get why use these helpers for nodes that aren't a Function Pointer type syntax. We can simply call the API and assert its result by using ToTestDisplayString. I think this makes the test very hard to follow and also makes the test to do a lot of unnecessary work.

333fred replied:
In the next iteration the helper will verify both type info and symbol info.

---
In reply to: [428917310](https://github.com/dotnet/roslyn/pull/44436#discussion_r428917310) [](ancestors = 428917310)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+            verifyStringReturn(assignment.Right, useConvertedType: true);
+
+            InvocationExpressionSyntax invocationExpressionSyntax = mDeclSyntax.DescendantNodes().OfType<InvocationExpressionSyntax>().Single();
+            verifyStringReturn(invocationExpressionSyntax.Expression, useConvertedType: false);
```
>invocationExpressionSyntax [](start = 31, length = 26)

I think it is interesting to testbehavior of APIs for ```invocationExpressionSyntax``` as well.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+        }
+
+        [Fact]
+        public void PublicApi_GetSymbolInfo()
```
>PublicApi_GetSymbolInfo [](start = 20, length = 23)

It feels like covering GetTypeInfo for error cases is interesting too.

AlekseyTs commented on src/Compilers/Test/Utilities/CSharp/FunctionPointerUtilities.cs:
```diff
@@ -128,6 +130,42 @@ static void verifyParameter(ParameterSymbol symbol, MethodSymbol containing)
+            params Action<ITypeSymbol>[] paramVerifiers)
+        {
+            var typeInfo = model.GetTypeInfo(syntax);
+            var typeSymbol = (IFunctionPointerTypeSymbol)(useConvertedType ? typeInfo.ConvertedType : typeInfo.Type)!;
```
>typeSymbol [](start = 16, length = 10)

Instead of taking a bunch of delegates, couldn't we just assert ```typeSymbol.ToTestDisplayString()```?

333fred replied:
The intent is to verify _both_ the full function pointer type, and the type we get from calling `GetTypeInfo` on the individual parts of the syntax. I felt this was best achieved by this pattern.

---
In reply to: [428922101](https://github.com/dotnet/roslyn/pull/44436#discussion_r428922101) [](ancestors = 428922101)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+            var model = comp.GetSemanticModel(syntaxTree);
+
+            var methodDecls = syntaxTree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().ToList();
+            var m2DeclSyntax = methodDecls.Skip(1).First();
```
>.Skip(1).First() [](start = 42, length = 16)

Isn't this a list? Can't we just index into it?

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+            var syntaxTree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(syntaxTree);
+
+            var methodDecls = syntaxTree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().ToList();
```
>ToList [](start = 103, length = 6)

```ToArray```? Are we planning to mutate the list (add/remove)?

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+            var symbolInfo = model.GetSymbolInfo(decls[0].Initializer!.Value);
+            Assert.Equal("System.String C.M1(System.Int32 i)", symbolInfo.Symbol.ToTestDisplayString());
+
+            symbolInfo = model.GetSymbolInfo(decls[1].Initializer!.Value);
```
>decls[1].Initializer!.Value [](start = 45, length = 27)

It would be interesting to test behavior for the operand of this node.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+            var decls = m2DeclSyntax.DescendantNodes().OfType<VariableDeclaratorSyntax>().ToList();
+
+            var symbolInfo = model.GetSymbolInfo(decls[0].Initializer!.Value);
+            Assert.Equal("System.String C.M1(System.Int32 i)", symbolInfo.Symbol.ToTestDisplayString());
```
>symbolInfo [](start = 63, length = 10)

What is the CandidateReason here?

333fred replied:
`VerifyFunctionPointerSemanticInfo` will verify this now.

---
In reply to: [428930132](https://github.com/dotnet/roslyn/pull/44436#discussion_r428930132) [](ancestors = 428930132)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+            symbolInfo = model.GetSymbolInfo(m3DeclSyntax.DescendantNodes().OfType<InvocationExpressionSyntax>().Single());
+            Assert.Empty(symbolInfo.CandidateSymbols);
+            Assert.Equal("delegate*<System.Int32, System.String>", symbolInfo.Symbol.ToTestDisplayString());
+        }
```
>} [](start = 8, length = 1)

For both tests, it is hard to tell whether all interesting scenarios are covered. Instead of trying to create new tests focused solely on SemanticModel, I would modify a set of existing tests (a set of tests covering the full breadth of "unique" success and error scenarios). I would also cover the same set of scenarios for IOperation.

333fred replied:
I've gone through and added a bunch more testing to existing scenario tests. I haven't done every single variation as I didn't think there were meaningful differences from a public API perspective, but I believe I've handled the majority of the interesting public api cases.

---
In reply to: [428933056](https://github.com/dotnet/roslyn/pull/44436#discussion_r428933056) [](ancestors = 428933056)

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/FunctionPointerOperations.cs:
```diff
@@ -0,0 +1,439 @@
+            var expectedOperationTree = @"
+IAddressOfOperation (OperationKind.AddressOf, Type: delegate*<System.Void>) (Syntax: '&M1')
+  Reference: 
+    IMethodReferenceOperation: delegate*<System.Void> (OperationKind.MethodReference, Type: null) (Syntax: 'M1')
```
>delegate*<System.Void> [](start = 31, length = 22)

This should be M1, I think.

AlekseyTs commented:
Done with review pass (iteration 4)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+            symbolInfo = model.GetSymbolInfo(m3DeclSyntax.DescendantNodes().OfType<InvocationExpressionSyntax>().Single());
+            Assert.Empty(symbolInfo.CandidateSymbols);
+            Assert.Equal("delegate*<System.Int32, System.String>", symbolInfo.Symbol.ToTestDisplayString());
+        }
```
>} [](start = 8, length = 1)

Do we have coverage for SemanticModel around conversions between FP types?

333fred replied:
Added to the existing conversions tests.

---
In reply to: [428938198](https://github.com/dotnet/roslyn/pull/44436#discussion_r428938198) [](ancestors = 428938198)

333fred commented:
@AlekseyTs address feedback.

---
In reply to: [632365282](https://github.com/dotnet/roslyn/pull/44436#issuecomment-632365282) [](ancestors = 632365282)

AlekseyTs commented on eng/build.sh:
```diff
@@ -235,6 +240,14 @@ function BuildSolution {
+    export ROSLYN_TEST_IOPERATION="true"
+
+    if [[ "$test_mono" != true && "$test_core_clr" != true ]]; then
+      test_core_clr=true
```
>test_core_clr=true [](start = 6, length = 18)

It doesn't look like any other option implies this.

333fred replied:
This mirrors `build.ps1` behavior.

---
In reply to: [430712569](https://github.com/dotnet/roslyn/pull/44436#discussion_r430712569) [](ancestors = 430712569)

AlekseyTs replied:
> This mirrors build.ps1 behavior.

Does this mean one cannot test IOperations for desktop?

---
In reply to: [430718240](https://github.com/dotnet/roslyn/pull/44436#discussion_r430718240) [](ancestors = 430718240,430712569)

333fred replied:
You can. On windows, this flag targets desktop by default, which is overridable. We can do that because the framework is installed on all platforms. For Mac/Linux, the default target framework is core clr, which is also overridable. We have no requirement that users install mono in order to do development on these platforms, and indeed I have not installed it. I did an initial smoke test of the IOperation flag on Linux, then went on my windows box to run the full set of tests (including IDE tests).

---
In reply to: [430719605](https://github.com/dotnet/roslyn/pull/44436#discussion_r430719605) [](ancestors = 430719605,430718240,430712569)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationNodes.cs:
```diff
@@ -47,8 +47,8 @@ internal abstract class LazyNoneOperation : BaseNoneOperation
 
-        public LazyNoneOperation(SemanticModel semanticModel, SyntaxNode node, Optional<object> constantValue, bool isImplicit) :
-            base(semanticModel, node, constantValue: constantValue, isImplicit: isImplicit)
+        public LazyNoneOperation(SemanticModel semanticModel, SyntaxNode node, Optional<object> constantValue, bool isImplicit, ITypeSymbol type = null) :
```
>ITypeSymbol type = null [](start = 128, length = 23)

It looks like the ```NoneOperation``` and clonner should be adjusted as well. It would be more robust to make this parameter required.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationNodes.cs:
```diff
@@ -16,8 +16,8 @@ namespace Microsoft.CodeAnalysis.Operations
     {
-        protected BaseNoneOperation(SemanticModel semanticModel, SyntaxNode syntax, Optional<object> constantValue, bool isImplicit) :
-            base(OperationKind.None, semanticModel, syntax, type: null, constantValue, isImplicit)
+        protected BaseNoneOperation(SemanticModel semanticModel, SyntaxNode syntax, Optional<object> constantValue, bool isImplicit, ITypeSymbol type = null) :
```
>ITypeSymbol type = null [](start = 133, length = 23)

I think this should become a required parameter now.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -1787,6 +1787,24 @@ private bool IsInTypeofExpression(int position)
+        /// has the information for the full member list. We therefore need to treat the BoundMethodGroup as the lowest bound
+        /// node here.
+        /// </summary>
+        private BoundExpression AdjustLowestBoundNodeForFunctionPointerConversion(BoundExpression expr)
```
Do we end up with BoundConversion and BoundMethodGroup linked to the same syntax node? It feels like conversion should be linked to the "address of" and the method group should be linked to its operand. Am I misunderstanding the scenario?

333fred replied:
No, it's the exact opposite. We want to pull the symbols out of the `BoundMethodGroup` that underlies the `BoundConversion` when we call `GetSymbolInfo` or `GetMembers` on `&M`. However, the highest _and_ lowest `BoundNode` for `&M` is the `BoundConversion`: the `BoundMethodGroup` that has the full symbol list we need is associated with the `M` syntax. So we need to grab that node and use it for the lowest node. I can expand the example in this comment.

---
In reply to: [430726079](https://github.com/dotnet/roslyn/pull/44436#discussion_r430726079) [](ancestors = 430726079)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -1820,6 +1838,8 @@ private bool IsInTypeofExpression(int position)
                     return SymbolInfo.None;
             }
 
+            boundExpr = AdjustLowestBoundNodeForFunctionPointerConversion(boundExpr);
```
>boundExpr = AdjustLowestBoundNodeForFunctionPointerConversion(boundExpr); [](start = 12, length = 73)

It feels like for a successful method group to FP conversion we should be able to get the best candidate that was picked. Are we able to achieve this? It looks like we are always downgrading to a method group here.

333fred replied:
Yes. The highest `BoundNode` will be a `BoundConversion`, which will have the correct method selected if overload resolution was successful, and that will be used for the "successful" lookup below.

---
In reply to: [430729125](https://github.com/dotnet/roslyn/pull/44436#discussion_r430729125) [](ancestors = 430729125)

333fred replied:
Based on what we do for invocations, I'll remove special handling of this case.

---
In reply to: [430732394](https://github.com/dotnet/roslyn/pull/44436#discussion_r430732394) [](ancestors = 430732394,430729125)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -900,6 +940,19 @@ private IOperation CreateBoundConversionOperation(BoundConversion boundConversio
+                if (boundConversion is { Type: FunctionPointerTypeSymbol { Signature: var signature }, Syntax: var expression })
+                {
+                    Debug.Assert(boundConversion.Conversion.MethodSymbol is object);
+                    SyntaxNode referenceSyntax = (expression is PrefixUnaryExpressionSyntax { RawKind: (int)SyntaxKind.AddressOfExpression, Operand: var operand }) ? operand : expression;
```
>referenceSyntax [](start = 31, length = 15)

It looks like the local is not used.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs:
```diff
@@ -19,8 +17,8 @@ internal sealed class CSharpLazyNoneOperation : LazyNoneOperation
 
-        public CSharpLazyNoneOperation(CSharpOperationFactory operationFactory, BoundNode boundNode, SemanticModel semanticModel, SyntaxNode node, Optional<object> constantValue, bool isImplicit) :
-            base(semanticModel, node, constantValue: constantValue, isImplicit: isImplicit)
+        public CSharpLazyNoneOperation(CSharpOperationFactory operationFactory, BoundNode boundNode, SemanticModel semanticModel, SyntaxNode node, Optional<object> constantValue, bool isImplicit, ITypeSymbol type = null) :
```
> = null [](start = 212, length = 7)

Remove default value?

AlekseyTs commented on src/Compilers/Test/Utilities/CSharp/CompilationTestUtils.cs:
```diff
@@ -266,6 +266,17 @@ public static SemanticInfoSummary GetSemanticInfoSummary(this SemanticModel sema
                 summary.ImplicitConversion = semanticModel.GetConversion(initializer);
                 summary.MemberGroup = semanticModel.GetMemberGroup(initializer);
             }
+            else if (node is PatternSyntax pattern)
```
>else if (node is PatternSyntax pattern) [](start = 12, length = 39)

Is this change related to the goal of this PR?

333fred replied:
Yes. One of my added tests is getting the semantic information off a PatternSyntax that's a function pointer type. It will crash without this change.

---
In reply to: [430759883](https://github.com/dotnet/roslyn/pull/44436#discussion_r430759883) [](ancestors = 430759883)

AlekseyTs commented on src/Compilers/Test/Utilities/CSharp/FunctionPointerUtilities.cs:
```diff
@@ -128,6 +131,134 @@ static void verifyParameter(ParameterSymbol symbol, MethodSymbol containing)
+                var returnParam = paramSyntaxes[^1].Type;
+                var returnSemanticInfo = model.GetSemanticInfoSummary(returnParam!);
+
+                Assert.Equal(signature.ReturnType, returnSemanticInfo.Type, SymbolEqualityComparer.Default);
```
>Assert.Equal(signature.ReturnType, returnSemanticInfo.Type, SymbolEqualityComparer.Default); [](start = 16, length = 92)

It looks like there is fair amount of code duplication (with the body of the loop above) from here to the end of the function.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -2131,6 +2162,7 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
         {
             if (lowestBoundNode is BoundExpression boundExpr)
             {
+                boundExpr = AdjustLowestBoundNodeForFunctionPointerConversion(boundExpr);
```
>boundExpr = AdjustLowestBoundNodeForFunctionPointerConversion(boundExpr); [](start = 16, length = 73)

It looks like based on our offline conversation this special case is not needed too.

333fred replied:
Yep, I've already removed it locally.

---
In reply to: [430767234](https://github.com/dotnet/roslyn/pull/44436#discussion_r430767234) [](ancestors = 430767234)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3169,6 +3170,37 @@ .maxstack  2
+
+            AssertEx.Equal(expectedMembers, model.GetMemberGroup(addressOfs[1]).Select(m => m.ToTestDisplayString(includeNonNullable: false)));
+            AssertEx.Equal(expectedMembers, model.GetMemberGroup(addressOfs[1].Operand).Select(m => m.ToTestDisplayString(includeNonNullable: false)));
         }
```
>} [](start = 8, length = 1)

Is it interesting to test IOperation for this scenario, or is it covered elsewhere?

333fred replied:
I can add it, but I didn't consider to to be different in a meaningful sense from a regular address of as covered by the tests in `FunctionPointerOperations.cs` (since the real difference is in ensuring that `GetMemberGroup` behaves expectedly, and we don't have an IOperation equivalent for that).

---
In reply to: [430770434](https://github.com/dotnet/roslyn/pull/44436#discussion_r430770434) [](ancestors = 430770434)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3194,6 +3226,22 @@ unsafe static void Main()
+
+            AssertEx.Equal(expectedSymbolCandidates, model.GetMemberGroup(addressOf).Select(m => m.ToTestDisplayString(includeNonNullable: false)));
+            AssertEx.Equal(expectedSymbolCandidates, model.GetMemberGroup(addressOf.Operand).Select(m => m.ToTestDisplayString(includeNonNullable: false)));
         }
```
>} [](start = 8, length = 1)

Is it interesting to test IOperation for this scenario, or is it covered elsewhere?

333fred replied:
Same comment: I didn't consider it particularly different from the failed addressof versions in `FunctionPointerOperations.cs`. I can add verification here as well if you'd prefer though.

---
In reply to: [430770526](https://github.com/dotnet/roslyn/pull/44436#discussion_r430770526) [](ancestors = 430770526)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3893,6 +4044,31 @@ static unsafe void M2(delegate*<void> b)
+            var variableDeclaratorSyntax = syntaxTree.GetRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>().Single();
+
+            var methodGroup1 = variableDeclaratorSyntax.Initializer!.Value;
+            FunctionPointerUtilities.VerifyFunctionPointerSemanticInfo(model, methodGroup1,
```
>VerifyFunctionPointerSemanticInfo [](start = 37, length = 33)

Consider testing GetMemberGroup API for this node.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -2104,6 +2124,17 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
+                    // we should reflect that in the types and conversion we return.
+                    convertedType = type;
+                    convertedNullability = nullability;
+                    conversion = new Conversion(ConversionKind.MethodGroup, symbol, isExtensionMethod: false);
```
>conversion = new Conversion(ConversionKind.MethodGroup, symbol, isExtensionMethod: false); [](start = 20, length = 90)

Could you point me to a test that covers this code path?

333fred replied:
One example would be `PublicApi_SemanticInfo01` (`FunctionPointerTypeSymbolTests.cs` line 1585). Basically, we want to make `&M` represent the language semantic here: it's a typeless expression that can be target-typed via an implicit MethodGroup conversion.

---
In reply to: [430778595](https://github.com/dotnet/roslyn/pull/44436#discussion_r430778595) [](ancestors = 430778595)

AlekseyTs replied:
> One example would be PublicApi_SemanticInfo01 (FunctionPointerTypeSymbolTests.cs line 1585). Basically, we want to make &M represent the language semantic here: it's a typeless expression that can be target-typed via an implicit MethodGroup conversion.

It doesn't look like we are asserting conversion information there.

---
In reply to: [430781925](https://github.com/dotnet/roslyn/pull/44436#discussion_r430781925) [](ancestors = 430781925,430778595)

333fred replied:
Added to `VerifyFunctionPointerSemanticInfo` for all address of expressions.

---
In reply to: [430850328](https://github.com/dotnet/roslyn/pull/44436#discussion_r430850328) [](ancestors = 430850328,430781925,430778595)

AlekseyTs commented:
Done with review pass (iteration 7)

333fred commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3194,6 +3226,22 @@ unsafe static void Main()
+            AssertEx.Equal(expectedSymbolCandidates, model.GetMemberGroup(addressOf.Operand).Select(m => m.ToTestDisplayString(includeNonNullable: false)));
         }
 
         [Fact]
```
I've removed these asserts as `VerifyFunctionPointerSemantics` now handles the `AddressOf` case.

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [634353435](https://github.com/dotnet/roslyn/pull/44436#issuecomment-634353435) [](ancestors = 634353435)

AlekseyTs commented:
Done with review pass (iteration 8)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [634421183](https://github.com/dotnet/roslyn/pull/44436#issuecomment-634421183) [](ancestors = 634421183)

cston commented on src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:
```diff
@@ -258,6 +258,22 @@ public static ImmutableArray<byte> ToImmutable(this MemoryStream stream)
             }
         }
 
+        public static ImmutableArray<TResult> ZipAsArray<T1, T2, TArg, TResult>(this ImmutableArray<T1> self, ImmutableArray<T2> other, TArg arg, Func<T1, T2, TArg, int, TResult> map)
```
>TArg, int [](start = 159, length = 9)

Consider switching `int` and `TArg` parameters to match `SelectAsArray`.

cston replied:
>ZipAsArray [](start = 46, length = 10)

Please consider adding dedicated tests. See `ImmutableArrayExtensionsTests.ZipAsArray`.

cston commented on src/Compilers/Core/Portable/Symbols/IFunctionPointerTypeSymbol.cs:
```diff
@@ -5,10 +5,20 @@
-    // PROTOTYPE(func-ptr): Document
+    /// <summary>
+    /// Represents a function pointer type such as "delegate*&lt;void&gt;". Pointer types
+    /// are used only in unsafe code.
```
Is this statement necessary (or correct)? Won't we expose an `IFunctionPointerTypeSymbol` for uses in code not marked `unsafe`?

333fred replied:
It's copied from IPointerTypeSymbol. I can remove it though.

---
In reply to: [431341498](https://github.com/dotnet/roslyn/pull/44436#discussion_r431341498) [](ancestors = 431341498)

cston commented on src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:
```diff
@@ -258,6 +258,22 @@ public static ImmutableArray<byte> ToImmutable(this MemoryStream stream)
             }
         }
 
+        public static ImmutableArray<TResult> ZipAsArray<T1, T2, TArg, TResult>(this ImmutableArray<T1> self, ImmutableArray<T2> other, TArg arg, Func<T1, T2, TArg, int, TResult> map)
```
>this ImmutableArray<T1> self, ImmutableArray<T2> other, TArg arg, Func<T1, T2, TArg, int, TResult> map [](start = 80, length = 102)

Please consider switching the order of `int` and `TArg` parameters to `map`, and the order of `arg` and `map` parameters to method, to match other extension methods:
```
..., Func<T1, T2, int, TArg, TResult> map, TArg arg);
```

333fred replied:
I'll switch the order in the `map` parameter, but I'm not swapping `map` and `arg`. The existing overloads are painful to use because, when naturally typing, no type information will have been inferred for `TArg` when creating a lambda for `map`. At some point I'd actually like to go change all the rest of them to put `arg` first because it makes intellisense behave significantly better.

---
In reply to: [431355574](https://github.com/dotnet/roslyn/pull/44436#discussion_r431355574) [](ancestors = 431355574)

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -2104,6 +2104,17 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
+                    // we should reflect that in the types and conversion we return.
+                    convertedType = type;
+                    convertedNullability = nullability;
+                    conversion = new Conversion(ConversionKind.MethodGroup, symbol, isExtensionMethod: false);
```
>new Conversion(ConversionKind.MethodGroup, symbol, isExtensionMethod: false) [](start = 33, length = 76)

Is this different from `boundExpr.Conversion`?

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -3223,6 +3234,28 @@ public ILocalSymbol GetDeclaredSymbol(CatchDeclarationSyntax catchDeclaration, C
+                    {
+                        // We try to match the results given for a similar piece of syntax here: bad invocations. A
+                        // BoundUnconvertedAddressOfOperator represents this syntax: &M
+                        // Similarly, a BoudCall for a bad invocation represents this syntax: M(args)
```
>BoudCall [](start = 40, length = 8)

Typo.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -3223,6 +3234,28 @@ public ILocalSymbol GetDeclaredSymbol(CatchDeclarationSyntax catchDeclaration, C
+
+                case BoundKind.UnconvertedAddressOfOperator:
+                    {
+                        // We try to match the results given for a similar piece of syntax here: bad invocations. A
```
> A [](start = 113, length = 2)

Minor: Consider moving to the next line for readability.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -3223,6 +3234,28 @@ public ILocalSymbol GetDeclaredSymbol(CatchDeclarationSyntax catchDeclaration, C
+                        // We try to match the results given for a similar piece of syntax here: bad invocations. A
+                        // BoundUnconvertedAddressOfOperator represents this syntax: &M
+                        // Similarly, a BoudCall for a bad invocation represents this syntax: M(args)
+                        // Calling GetSymbolInfo on the syntax will return a list of candidate symbols that were
```
>a list [](start = 75, length = 6)

Minor: Consider using "array" here and in line below instead of "list".

cston commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -434,6 +438,42 @@ private IOperation CreateBoundCallOperation(BoundCall boundCall)
+                    boundFunctionPointerInvocation.Syntax,
+                    boundFunctionPointerInvocation.Type.GetPublicSymbol(),
+                    ConvertToOptional(boundFunctionPointerInvocation.ConstantValue),
+                    isImplicit: boundFunctionPointerInvocation.WasCompilerGenerated);
```
Consider extracting locals for each of these, for use here and below.

cston commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -900,6 +940,18 @@ private IOperation CreateBoundConversionOperation(BoundConversion boundConversio
             BoundExpression boundOperand = boundConversion.Operand;
             if (boundConversion.ConversionKind == CSharp.ConversionKind.MethodGroup)
             {
+                if (boundConversion is { Type: FunctionPointerTypeSymbol { Signature: var signature }, Syntax: var expression })
```
>{ Signature: var signature }, Syntax: var expression } [](start = 73, length = 54)

Are these variable declarations needed? `signature` is unused, and the `boundConversion.Syntax` can be used instead of `expression` (in the same way that `boundConversion.Operand` and `boundConversion.SymbolOpt` are).

cston commented:
>                 Optional<object> constantValue = ConvertToOptional(boundConversion.ConstantValue);

Consider moving these ahead of `if (boundConversion is ...)` so these values can be used in that `if` body as well.

---
Refers to: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:960 in 1edbf72. [](commit_id = 1edbf726a67140b340599ce956ced0f1d75d2317, deletion_comment = False)

cston commented on src/Compilers/CSharp/Portable/Symbols/SymbolExtensions.cs:
```diff
@@ -10,6 +10,7 @@
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
 using System.Linq;
+using Microsoft.CodeAnalysis.PooledObjects;
```
>Microsoft.CodeAnalysis.PooledObjects [](start = 6, length = 36)

Unused?

cston commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -92,6 +92,26 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
+            CSharpCompilation compilation)
+        {
+            return new FunctionPointerMethodSymbol(
+                // https://github.com/dotnet/roslyn/issues/39865 allow setting this in creation
```
>// https://github.com/dotnet/roslyn/issues/39865 allow setting this in creation [](start = 16, length = 79)

Does this refer to allowing `CallingConvention` to be settable? If so, does the comment belong on the public API instead?

cston commented on src/Compilers/CSharp/Test/IOperation/IOperation/FunctionPointerOperations.cs:
```diff
@@ -0,0 +1,538 @@
+        }
+
+        [Fact]
+        public void FunctionPointerLoad_UnknownMethod()
```
>FunctionPointerLoad_UnknownMethod [](start = 20, length = 33)

Is this different from `FunctionPointerLoad_NonExistantMethod`?

cston commented on src/Compilers/CSharp/Test/IOperation/IOperation/FunctionPointerOperations.cs:
```diff
@@ -0,0 +1,538 @@
+        }
+
+        [Fact]
+        public void FunctionPointerLoad_InCFG()
```
>Load [](start = 35, length = 4)

"Assignment"?

333fred replied:
No, the interesting thing I'm testing here is the address of. I'll use that in the name.

---
In reply to: [431389332](https://github.com/dotnet/roslyn/pull/44436#discussion_r431389332) [](ancestors = 431389332)

cston commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:
```diff
@@ -2727,6 +2727,14 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Throw New NotSupportedException(VBResources.ThereAreNoPointerTypesInVB)
         End Function
 
+        Protected Overrides Function CommonCreateFunctionPointerTypeSymbol(
```
>CommonCreateFunctionPointerTypeSymbol [](start = 37, length = 37)

Please test this method. Perhaps simply update `CompilationAPITests.CompilationStackOverflow()`.

333fred replied:
I'm going to rename that test...

---
In reply to: [431405059](https://github.com/dotnet/roslyn/pull/44436#discussion_r431405059) [](ancestors = 431405059)

cston commented on src/Compilers/Core/CodeAnalysisTest/Collections/ImmutableArrayExtensionsTests.cs:
```diff
@@ -421,6 +421,38 @@ public void ZipAsArray()
         }
 
+        [Fact]
+        public void ZipAsArrayWithPredicate()
```
>WithPredicate [](start = 30, length = 13)

There's no predicate. Perhaps "WithIndex"?

jasonmalinowski commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -939,6 +939,28 @@ public IPointerTypeSymbol CreatePointerTypeSymbol(ITypeSymbol pointedAtType)
+        ///  * parameterTypes and parameterRefKinds do not have the same length.
+        /// </exception>
+        /// <exception cref="ArgumentNullException">
+        /// If returnType is <see langword="null"/>, or if parameterTypes or parameterRefKinds are default,
+        /// or if any of the types in parameterTypes are null.</exception>
```
The expectation is the return type is System.Void?

333fred replied:
Yep.
